impeller/renderer/backend/vulkan/command_pool_vk.h

```diff
@@ -5,109 +5,126 @@
-  std::weak_ptr<const DeviceHolder> device_holder_;
-  vk::UniqueCommandPool graphics_pool_;
-  Mutex buffers_to_collect_mutex_;
-  std::vector<vk::UniqueCommandBuffer> buffers_to_collect_
-      IPLR_GUARDED_BY(buffers_to_collect_mutex_);
-  std::vector<vk::UniqueCommandBuffer> recycled_buffers_;
-  bool is_valid_ = false;
-
-  /// @brief      Resets, releasing all |vk::CommandBuffer|s.
+  FML_DISALLOW_COPY_AND_ASSIGN(CommandPoolVK);
+
+  vk::UniqueCommandPool pool_;
+  std::weak_ptr<ContextVK>& context_;
+
+  // Used to retain a reference on these until the pool is reset.
+  std::vector<vk::UniqueCommandBuffer> collected_buffers_;
+};
+
+//------------------------------------------------------------------------------
+/// @brief      Creates and manages the lifecycle of |vk::CommandPool| objects.
+///
+/// A |vk::CommandPool| is expensive to create and reset. This class manages
+/// the lifecycle of |vk::CommandPool| objects by creating and recycling them;
+/// or in other words, a pool for command pools.
+///
+/// A single instance should be created per |ContextVK|.
+///
+/// Every "frame", a single |CommandPoolResourceVk| is made available for each
+/// thread that calls |Get|. After calling |Recycle|, the current thread's pool
+/// is moved to a background thread, reset, and made available for the next time
+/// |Get| is called and needs to create a command pool.
+///
+/// @note       This class is thread-safe.
+///
+/// @see        |vk::CommandPoolResourceVk|
+/// @see        |ContextVK|
+/// @see
+/// https://arm-software.github.io/vulkan_best_practice_for_mobile_developers/samples/performance/command_buffer_usage/command_buffer_usage_tutorial.html
+class CommandPoolRecyclerVK final
+    : public std::enable_shared_from_this<CommandPoolRecyclerVK> {
+ public:
+  ~CommandPoolRecyclerVK();
+
+  /// @brief      Creates a recycler for the given |ContextVK|.
   ///
-  /// @note       "All" includes active and recycled buffers.
-  void Reset();
+  /// @param[in]  context The context to create the recycler for.
+  explicit CommandPoolRecyclerVK(std::weak_ptr<ContextVK> context)
+      : context_(std::move(context)) {}
 
-  explicit CommandPoolVK(const ContextVK* context);
+  /// @brief      Gets a command pool for the current thread.
+  ///
+  /// @warning    Returns a |nullptr| if a pool could not be created.
+  std::shared_ptr<CommandPoolVK> Get();
 
-  /// @brief      Collects buffers for recycling if able.
+  /// @brief      Returns a command pool to be reset on a background thread.
   ///
-  /// If any buffers have been returned through |CollectGraphicsCommandBuffer|,
-  /// then they are reset and made available to future calls to
-  /// |CreateGraphicsCommandBuffer|.
+  /// @param[in]  pool The pool to recycle.
+  void Reclaim(vk::UniqueCommandPool&& pool);
+
+  /// @brief      Clears all recycled command pools to let them be reclaimed.
+  void Recycle();
+
+ private:
+  FML_DISALLOW_COPY_AND_ASSIGN(CommandPoolRecyclerVK);
```

Comment:
Minor Nit: These go last. The general order is. Private struct definitions -> typedefs -> private ivars -> private methods -> macros.

Reply:
Done.